lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -797,7 +944,201 @@ func (d *zfs) CreateVolumeFromMigration(vol Volume, conn io.ReadWriteCloser, vol
 
 // RefreshVolume updates an existing volume to match the state of another.
 func (d *zfs) RefreshVolume(vol Volume, srcVol Volume, srcSnapshots []Volume, op *operations.Operation) error {
-	return genericVFSCopyVolume(d, nil, vol, srcVol, srcSnapshots, true, op)
+	// Get target snapshots
+	targetSnapshots, err := vol.Snapshots(op)
+	if err != nil {
+		return fmt.Errorf("Failed to get target snapshots: %w", err)
+	}
+
+	srcSnapshotsAll, err := srcVol.Snapshots(op)
+	if err != nil {
+		return fmt.Errorf("Failed to get source snapshots: %w", err)
+	}
+
+	// If there are no target or source snapshots, perform a simple copy using zfs.
+	// We cannot use generic vfs volume copy here, as zfs will complain if a generic
+	// copy/refresh is followed by an optimized refresh.
+	if len(targetSnapshots) == 0 || len(srcSnapshotsAll) == 0 {
+		err = d.DeleteVolume(vol, op)
+		if err != nil {
+			return err
+		}
+
+		return d.CreateVolumeFromCopy(vol, srcVol, len(srcSnapshots) > 0, op)
+	}
+
+	transfer := func(src Volume, target Volume, origin Volume) error {
+		var sender *exec.Cmd
+
+		receiver := exec.Command("zfs", "receive", d.dataset(target, false))
+
+		if origin.Name() != src.Name() {
+			sender = exec.Command("zfs", "send", "-i", d.dataset(origin, false), d.dataset(src, false))
+		} else {
+			sender = exec.Command("zfs", "send", d.dataset(src, false))
+		}
+
+		var senderErrBuf bytes.Buffer
+		var receiverErrBuf bytes.Buffer
+
+		// Configure the pipes.
+		sender.Stderr = &senderErrBuf
+		receiver.Stdin, _ = sender.StdoutPipe()
+		receiver.Stdout = os.Stdout
+		receiver.Stderr = &receiverErrBuf
+
+		// Run the transfer.
+		err := receiver.Start()
+		if err != nil {
+			return fmt.Errorf("Failed to receive stream: %w", err)
+		}
+
+		err = sender.Run()
+		if err != nil {
+			// This removes any newlines in the error message.
+			msg := strings.ReplaceAll(senderErrBuf.String(), "\n", " ")
+
+			return fmt.Errorf("Failed to send stream %q: %s: %w", sender.String(), msg, err)
+		}
+
+		err = receiver.Wait()
+		if err != nil {
+			// This removes any newlines in the error message.
+			msg := strings.ReplaceAll(receiverErrBuf.String(), "\n", " ")
+
+			if strings.Contains(msg, "does not match incremental source") {
```

Comment:
Thanks. We might be pushing our luck here, but does the command return a specific exit code when this happens? It would be great not to have to rely on the text matching, but if it doesn't then we'll have to use this approach.

Reply:
Unfortunately, it's all exit code 1.